src/v/cluster/scheduling/partition_allocator.cc

```diff
@@ -81,6 +81,92 @@ allocation_constraints partition_allocator::default_constraints() {
+    // the total memory space for partitions. Instead of dividing the whole
+    // memory space by `topic_memory_per_partition` we now only do that with the
+    // reserved value.
+    //
+    // Because of the latter making this change isn't trivial. Users might have
+    // lowered the `topic_memory_per_partition` property to gain higher
+    // partition density. With only using the reserved memory space when
+    // checking memory limits this might lead to now actually having less
+    // partition density. Hence, we need to guess whether the property was set
+    // before this change or after. If it was set before we just use the old
+    // rules (while still reserving a part of the memory) when checking memory
+    // limits to not break existing behaviour.
+    //
+    // Note that if we get this guess wrong that isn't immediately fatal. This
+    // only becomes active when creating new topics or modifying existing
+    // topics. At that point users can increase the memory reservation if
+    // needed.
+
+    // not overriden so definitely memory group aware
+    if (!config::shard_local_cfg().topic_memory_per_partition.is_overriden()) {
+        return true;
+    }
+
+    auto value = config::shard_local_cfg().topic_memory_per_partition.value();
+
+    // Previous default was 4MiB (and even higher before that). New one is more
+    // than 10x smaller. We assume it's unlikely someone would have changed the
+    // value to be that much smaller and hence guess that all the values larger
+    // than 2 times the new default are old values. Equally in the new world
+    // it's unlikely anybody would increase the value (at all really).
+
+    return value < 2 * ORIGINAL_MEMORY_GROUP_AWARE_TMPP;
+}
+
+std::error_code partition_allocator::check_memory_limits(
+  const uint64_t new_partitions_replicas_requested,
+  uint64_t proposed_total_partitions,
+  uint64_t effective_cluster_memory) const {
+    // Refuse to create partitions that would violate the configured
+    // memory per partition.
+    auto memory_per_partition_replica
+      = config::shard_local_cfg().topic_memory_per_partition();
+
+    if (!memory_per_partition_replica.has_value()) {
+        return errc::success;
+    }
+
+    bool is_memory_aware = guess_is_memory_group_aware_memory_limit();
+
+    uint64_t memory_limit;
+    if (is_memory_aware) {
+        memory_limit = effective_cluster_memory
+                       * memory_groups().partitions_max_memory_share()
+                       / memory_per_partition_replica.value();
+    } else {
+        memory_limit = effective_cluster_memory
+                       / memory_per_partition_replica.value();
+    }
+
+    if (proposed_total_partitions > memory_limit) {
+        vlog(
+          clusterlog.warn,
+          "Refusing to create {} new partitions as total partition count "
+          "{} "
+          "would exceed memory limit {}",
```

Comment:
Please include in the error message more components of the math e.g., `memory_limit` and other values so we can reconstruct what happened w/o knowing the conifg values.

Reply:
(yes, I know this is c/p from elsewhere but I think this is even more important now)

Reply:
Added a few more. Let me know if you think anything is missing.